coordinator-common/src/main/java/org/apache/kafka/coordinator/common/runtime/CoordinatorRecordSerde.java

```diff
@@ -0,0 +1,122 @@
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kafka.coordinator.common.runtime;
+
+import org.apache.kafka.common.protocol.ApiMessage;
+import org.apache.kafka.common.protocol.ByteBufferAccessor;
+import org.apache.kafka.common.protocol.MessageUtil;
+import org.apache.kafka.server.common.ApiMessageAndVersion;
+
+import java.nio.BufferUnderflowException;
+import java.nio.ByteBuffer;
+
+/**
+ * Serializer/Deserializer for {@link CoordinatorRecord}. The format is defined below:
+ * <pre>
+ *     record_key   = [record_type key_message]
+ *     record_value = [value_version value_message]
+ *
+ *     record_type     : The record type is currently define as the version of the key
+ *                       {@link ApiMessageAndVersion} object.
+ *     key_message     : The serialized message of the key {@link ApiMessageAndVersion} object.
+ *     value_version   : The value version is currently define as the version of the value
+ *                       {@link ApiMessageAndVersion} object.
+ *     value_message   : The serialized message of the value {@link ApiMessageAndVersion} object.
+ * </pre>
+ */
+public abstract class CoordinatorRecordSerde implements Serializer<CoordinatorRecord>, Deserializer<CoordinatorRecord> {
+    @Override
+    public byte[] serializeKey(CoordinatorRecord record) {
+        // Record does not accept a null key.
+        return MessageUtil.toVersionPrefixedBytes(
+            record.key().version(),
+            record.key().message()
+        );
+    }
+
+    @Override
+    public byte[] serializeValue(CoordinatorRecord record) {
+        // Tombstone is represented with a null value.
+        if (record.value() == null) {
+            return null;
+        } else {
+            return MessageUtil.toVersionPrefixedBytes(
+                record.value().version(),
+                record.value().message()
+            );
+        }
+    }
+
+    @Override
+    public CoordinatorRecord deserialize(
+        ByteBuffer keyBuffer,
+        ByteBuffer valueBuffer
+    ) throws RuntimeException {
+        final short recordType = readVersion(keyBuffer, "key");
+        final ApiMessage keyMessage = apiMessageKeyFor(recordType);
+        readMessage(keyMessage, keyBuffer, recordType, "key");
+
+        if (valueBuffer == null) {
+            return new CoordinatorRecord(new ApiMessageAndVersion(keyMessage, recordType), null);
+        }
+
+        final ApiMessage valueMessage = apiMessageValueFor(recordType);
```

Comment:
I know the current value version is equal to key version, but we should pass `valueVersion` to avoid possible bugs in the future, right?

Reply:
@chia7712 
Not sure about this - with this change a lot of tests in `GroupCoordinatorRecordSerdeTest` will fail. 
IDK if those tests are deliberate. 
They pass differing values for versions of keys and values (`testDeserialize`). With the above change they will pass if the versions are modified as per `apiMessageFor(recordType)` implementation. That would need further analysis of `GroupCoordinatorRecordHelpers`.
I would be hesitant to make this change unless the original author @dajac could provide some input.

Reply:
ok, I gets confused by the `SHARE_UPDATE_RECORD_VALUE_VERSION`. Let me re-comment it again.

`final ApiMessage valueMessage = apiMessageValueFor(recordType);` is correct, as the type of value is associated to key, so they share the same `recordType`. that is good.

The code which confuses me is shown below. `SHARE_UPDATE_RECORD_VALUE_VERSION` is `1` and `ShareUpdateValue`'s version is (0, 0). I assume the version added to `ApiMessageAndVersion` is the protocol version. Please correct me if I'm wrong 
```java
            new ApiMessageAndVersion(new ShareUpdateValue()
                .setSnapshotEpoch(offsetData.snapshotEpoch())
                .setLeaderEpoch(offsetData.leaderEpoch())
                .setStartOffset(offsetData.startOffset())
                .setStateBatches(offsetData.stateBatches().stream()
                    .map(batch -> new ShareUpdateValue.StateBatch()
                        .setFirstOffset(batch.firstOffset())
                        .setLastOffset(batch.lastOffset())
                        .setDeliveryCount(batch.deliveryCount())
                        .setDeliveryState(batch.deliveryState()))
                    .collect(Collectors.toList())),
                ShareCoordinator.SHARE_UPDATE_RECORD_VALUE_VERSION)
```

Reply:
> Yes, we can keep this 0 - not necessary to use constant.

yes, that is what I expected :)

> The constant is needed in apiMessageValueFor impl in ShareCoordinatorRecordSerde.
The CoordinatorRecordSerde.deserialize method reads the key version and supplies the same while fetching the ApiMessage object to read the data into

thanks for the explanation!